src/app/extract_blocks/extract_blocks.ml

```diff
@@ -311,47 +330,63 @@ let main ~archive_uri ~start_state_hash_opt ~end_state_hash () =
-            [%log info]
-              "Querying for subchain to end block with given state hash" ;
-            query_db pool
-              ~f:(fun db ->
-                Sql.Subchain.start_from_unparented db ~end_state_hash )
-              ~item:"blocks starting from unparented"
-        | Some start_state_hash ->
-            [%log info]
-              "Querying for subchain from start block to end block with given \
-               state hashes" ;
-            query_db pool
-              ~f:(fun db ->
-                Sql.Subchain.start_from_specified db ~start_state_hash
-                  ~end_state_hash )
-              ~item:"blocks starting from specified"
+        if all_blocks then (
+          [%log info] "Querying for all blocks" ;
+          query_db pool
+            ~f:(fun db -> Sql.Subchain.all_blocks db)
+            ~item:"all blocks" )
+        else
+          match (start_state_hash_opt, end_state_hash_opt) with
+          | None, Some end_state_hash ->
+              [%log info]
+                "Querying for subchain to end block with given state hash" ;
+              query_db pool
+                ~f:(fun db ->
+                  Sql.Subchain.start_from_unparented db ~end_state_hash )
+                ~item:"blocks starting from unparented"
+          | Some start_state_hash, Some end_state_hash ->
+              [%log info]
+                "Querying for subchain from start block to end block with \
+                 given state hashes" ;
+              query_db pool
+                ~f:(fun db ->
+                  Sql.Subchain.start_from_specified db ~start_state_hash
+                    ~end_state_hash )
+                ~item:"blocks starting from specified"
+          | _ ->
+              (* unreachable *)
+              failwith "Unexpected flag combination"
       in
-      let end_block_found =
-        List.exists blocks ~f:(fun block ->
-            String.equal block.state_hash end_state_hash )
-      in
-      ( match start_state_hash_opt with
-      | Some state_hash ->
-          if
-            not
-              ( end_block_found
-              && List.exists blocks ~f:(fun block ->
-                     String.equal block.state_hash state_hash ) )
-          then (
-            [%log error]
-              "No subchain with given start and end state hashes available; \
-               try omitting the start state hash, to get a chain from an \
-               unparented block to the block with the end state hash" ;
-            Core.exit 1 )
+      ( match end_state_hash_opt with
       | None ->
-          if not end_block_found then (
-            [%log error]
-              "No subchain available from an unparented block (possibly the \
-               genesis block) to block with given end state hash" ;
-            Core.exit 1 ) ) ;
+          (* all blocks *)
+          ()
+      | Some end_state_hash -> (
+          (* some blocks, check whether we have the desired chain *)
```

Comment:
May I suggest to move this to the pattern match above when `Some end_state_hash` is true? It a bit confusing for me when there are different possible cases in multiple places